File: packages/control-contract/src/actor.ts
/**
 * Actor identity for mutations.
 *
 * Every mutation command includes an `Actor` that identifies who performed
 * the change. An actor is never inferred from a session ID or chat name; it
 * is always explicitly supplied by the authenticated caller.
 */

import type { TypedId } from "./ids.js";
import { OrganizationId } from "./ids.js";

/** A user identity scoped to an organization. Distinct from agent IDs. */
export type UserId = {
  readonly organization_id: TypedId;
  readonly value: string;
};

export function createUserId(organizationId: TypedId, value: string): UserId {
  return { organization_id: organizationId, value };
}

export type ActorKind = "user" | "agent" | "system" | "external";

/** Who performed a mutation. */
export type Actor =
  | {
      readonly kind: "user";
      readonly id: UserId;
      /** Display name at the time of the mutation (denormalized for audit). */
      readonly display_name: string;
    }
  | {
      readonly kind: "agent";
      readonly id: TypedId;
      /** The attempt that authorized this action, if applicable. */
      readonly attempt_id?: string;
    }
  | {
      readonly kind: "system";
      /** Free-form component name for diagnostics. */
      readonly component: string;
    }
  | {
      readonly kind: "external";
      readonly integration: string;
      readonly external_actor_id: string;
    };

export function actorKind(actor: Actor): ActorKind {
  return actor.kind;
}

export function actorOrganization(actor: Actor): TypedId | undefined {
  return actor.kind === "user" ? actor.id.organization_id : undefined;
}

/** Parse an unknown value into an Actor, rejecting malformed input. */
export function parseActor(input: unknown): Actor {
  if (typeof input !== "object" || input === null || Array.isArray(input)) {
    throw new Error("actor must be an object");
  }
  const candidate = input as { kind?: unknown };
  if (typeof candidate.kind !== "string") {
    throw new Error("actor missing 'kind' field");
  }
  switch (candidate.kind) {
    case "user": {
      const a = input as { id?: unknown; display_name?: unknown };
      if (typeof a.id !== "object" || a.id === null) {
        throw new Error("user actor requires 'id'");
      }
      const id = a.id as { organization_id?: unknown; value?: unknown };
      if (typeof id.value !== "string") throw new Error("user id requires 'value'");
      const organizationId = OrganizationId.parse(id.organization_id);
      if (typeof a.display_name !== "string") {
        throw new Error("user actor requires 'display_name'");
      }
      return {
        kind: "user",
        id: { organization_id: organizationId, value: id.value },
        display_name: a.display_name,
      };
    }
    case "agent": {
      const a = input as { id?: unknown; attempt_id?: unknown };
      if (typeof a.id !== "object" || a.id === null) {
        throw new Error("agent actor requires 'id'");
      }
      return {
        kind: "agent",
        id: a.id as TypedId,
        attempt_id:
          typeof a.attempt_id === "string" ? a.attempt_id : undefined,
      };
    }
    case "system": {
      const a = input as { component?: unknown };
      if (typeof a.component !== "string") {
        throw new Error("system actor requires 'component'");
      }
      return { kind: "system", component: a.component };
    }
    case "external": {
      const a = input as { integration?: unknown; external_actor_id?: unknown };
      if (typeof a.integration !== "string" || typeof a.external_actor_id !== "string") {
        throw new Error("external actor requires 'integration' and 'external_actor_id'");
      }
      return {
        kind: "external",
        integration: a.integration,
        external_actor_id: a.external_actor_id,
      };
    }
    default:
      throw new Error(`unknown actor kind: ${candidate.kind}`);
  }
}

File: src-tauri/crates/altai-control-protocol/src/actor.rs
//! Actor identity for mutations.
//!
//! Every mutation command includes an [`Actor`] that identifies who
//! performed the change. An actor is never inferred from a session ID or
//! chat name; it is always explicitly supplied by the authenticated caller.

use crate::id::{AgentInstanceId, OrganizationId};
use serde::{Deserialize, Serialize};
use std::fmt;

/// Who performed a mutation.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Actor {
    /// A human user.
    User {
        id: UserId,
        /// Display name at the time of the mutation (denormalized for audit).
        display_name: String,
    },
    /// An agent instance acting on behalf of the control plane.
    Agent {
        id: AgentInstanceId,
        /// The attempt that authorized this action, if applicable.
        attempt_id: Option<String>,
    },
    /// The control-plane system itself (e.g. recovery sweeps, scheduler).
    System {
        /// Free-form component name for diagnostics.
        component: String,
    },
    /// An external integration (e.g. GitHub webhook).
    External {
        integration: String,
        external_actor_id: String,
    },
}

/// The kind of actor, for quick filtering without matching the full enum.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ActorKind {
    User,
    Agent,
    System,
    External,
}

impl Actor {
    pub fn kind(&self) -> ActorKind {
        match self {
            Self::User { .. } => ActorKind::User,
            Self::Agent { .. } => ActorKind::Agent,
            Self::System { .. } => ActorKind::System,
            Self::External { .. } => ActorKind::External,
        }
    }

    /// The organization this actor belongs to, if known.
    pub fn organization(&self) -> Option<&OrganizationId> {
        match self {
            Self::User { id, .. } => Some(&id.organization_id),
            _ => None,
        }
    }
}

impl fmt::Display for Actor {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::User { id, .. } => write!(f, "user:{}", id.value),
            Self::Agent { id, .. } => write!(f, "agent:{}", id.value),
            Self::System { component } => write!(f, "system:{}", component),
            Self::External { integration, external_actor_id } => {
                write!(f, "external:{}/{}", integration, external_actor_id)
            }
        }
    }
}

/// A user identity scoped to an organization. Distinct from agent IDs.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct UserId {
    pub organization_id: OrganizationId,
    pub value: String,
}

impl UserId {
    pub fn new(organization_id: OrganizationId, value: impl Into<String>) -> Self {
        Self {
            organization_id,
            value: value.into(),
        }
    }
}

impl fmt::Display for UserId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "user:{}/{}", self.organization_id.value, self.value)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn actor_user_serializes() {
        let org = OrganizationId::new("test-org");
        let user = UserId::new(org, "alice");
        let actor = Actor::User {
            id: user,
            display_name: "Alice".to_string(),
        };
        let json = serde_json::to_string(&actor).unwrap();
        assert!(json.contains("\"kind\":\"user\""));
        assert!(json.contains("\"alice\""));
    }

    #[test]
    fn actor_agent_serializes() {
        let agent = AgentInstanceId::new("dev-agent-1");
        let actor = Actor::Agent {
            id: agent,
            attempt_id: Some("att_123".to_string()),
        };
        let json = serde_json::to_string(&actor).unwrap();
        assert!(json.contains("\"kind\":\"agent\""));
    }

    #[test]
    fn actor_system_serializes() {
        let actor = Actor::System {
            component: "recovery_sweep".to_string(),
        };
        let json = serde_json::to_string(&actor).unwrap();
        assert!(json.contains("\"kind\":\"system\""));
    }

    #[test]
    fn actor_kind_matches() {
        let org = OrganizationId::new("test-org");
        let actor = Actor::User {
            id: UserId::new(org, "bob"),
            display_name: "Bob".to_string(),
        };
        assert_eq!(actor.kind(), ActorKind::User);
    }
}
